mono/metadata/marshal.c

```diff
@@ -7755,30 +7840,26 @@ mono_marshal_emit_managed_wrapper (MonoMethodBuilder *mb, MonoMethodSignature *i
 		mono_mb_add_local (mb, sig->ret);
 	}
 
-	if (mono_threads_is_coop_enabled ()) {
-		/* local 4, the local to be used when calling the reset_blocking funcs */
-		/* tons of code hardcode 3 to be the return var */
-		coop_gc_var = mono_mb_add_local (mb, &mono_defaults.int_class->byval_arg);
-		/* local 5, the local used to get a stack address for suspend funcs */
-		coop_gc_dummy_local = mono_mb_add_local (mb, &mono_defaults.int_class->byval_arg);
-	}
+	/* local 4, the local to be used when calling the mono_marshal_thread_attach funcs */
+	/* tons of code hardcode 3 to be the return var */
+	cookie = mono_mb_add_local (mb, &mono_defaults.int_class->byval_arg);
+	/* local 5, the local used to get a stack address for suspend funcs */
+	dummy = mono_mb_add_local (mb, &mono_defaults.int_class->byval_arg);
 
 	mono_mb_emit_icon (mb, 0);
 	mono_mb_emit_stloc (mb, 2);
 
-	/*
-	 * Might need to attach the thread to the JIT or change the
-	 * domain for the callback.
-	 */
-	mono_mb_emit_byte (mb, MONO_CUSTOM_PREFIX);
-	mono_mb_emit_byte (mb, CEE_MONO_JIT_ATTACH);
-
-	if (mono_threads_is_coop_enabled ()) {
-		/* XXX can we merge reset_blocking_start with JIT_ATTACH above and save one call? */
-		mono_mb_emit_ldloc_addr (mb, coop_gc_dummy_local);
-		mono_mb_emit_icall (mb, mono_threads_reset_blocking_start);
-		mono_mb_emit_stloc (mb, coop_gc_var);
-	}
+	/* cookie = mono_marshal_threads_attach (mono_domain_get (), &dummy) */
+#if SIZEOF_VOID_P == 4
+	mono_mb_emit_byte (mb, CEE_LDC_I4);
+	mono_mb_emit_i4 (mb, aot ? -1 : (gint32)(gsize) mono_domain_get ());
```

Comment:
This is different than what the JIT does. The jit uses cfg->domain instead. Couldn't that be an issue?